esp-hal-common/src/uart.rs

```diff
@@ -1331,94 +1367,131 @@ mod asynch {
-            // The interrupt handler clears and disables all interrupts that occur
-            // and have their enable bit set
-            let still_enabled_interrupts =
-                UartInterrupt::from_bits_truncate(T::register_block().int_ena.read().bits());
-            // return only the interrupts that we have initially enabled
-            self.interrupts.difference(still_enabled_interrupts)
-        }
-
-        fn disable_interrupts(&self) {
-            // Because the interrupt handler only disables the interrupts that have occured,
-            // we have to disabled the remaining interrupts (that we have enabled ourselves)
-            T::register_block()
-                .int_ena
-                .modify(|r, w| unsafe { w.bits(r.bits() & self.interrupts.complement().bits()) });
+        fn event_bit_is_clear(&self) -> bool {
+            match self.event {
+                Event::TxDone => T::register_block()
+                    .int_ena
+                    .read()
+                    .tx_done_int_ena()
+                    .bit_is_clear(),
+                Event::TxFiFoEmpty => T::register_block()
+                    .int_ena
+                    .read()
+                    .txfifo_empty_int_ena()
+                    .bit_is_clear(),
+                Event::RxFifoFull => T::register_block()
+                    .int_ena
+                    .read()
+                    .rxfifo_full_int_ena()
+                    .bit_is_clear(),
+                Event::RxCmdCharDetected => T::register_block()
+                    .int_ena
+                    .read()
+                    .at_cmd_char_det_int_ena()
+                    .bit_is_clear(),
+                Event::RxFifoOvf => T::register_block()
+                    .int_ena
+                    .read()
+                    .rxfifo_ovf_int_ena()
+                    .bit_is_clear(),
+                Event::RxFifoTout => T::register_block()
+                    .int_ena
+                    .read()
+                    .rxfifo_tout_int_ena()
+                    .bit_is_clear(),
+            }
         }
     }
 
     impl<'d, T: Instance> core::future::Future for UartFuture<'d, T> {
-        type Output = UartInterrupt;
+        type Output = Event;
 
         fn poll(
             self: core::pin::Pin<&mut Self>,
             cx: &mut core::task::Context<'_>,
         ) -> core::task::Poll<Self::Output> {
             WAKERS[T::uart_number()].register(cx.waker());
-            let occured_interrupts = self.occured_interrupts();
-            if occured_interrupts.is_empty() {
-                Poll::Pending
+            if self.event_bit_is_clear() {
+                Poll::Ready(self.event)
             } else {
-                self.disable_interrupts();
-                Poll::Ready(occured_interrupts)
+                Poll::Pending
+            }
+        }
+    }
+
+    impl<'d, T: Instance> Drop for UartFuture<'d, T> {
+        fn drop(&mut self) {
```

Comment:
I like this a lot! Many of our async drivers don't disarm the ISR's on drop which can lead to some interesting behavior sometimes.